packages/@react-stately/tree/src/useTreeState.ts

```diff
@@ -78,13 +87,32 @@ export function useTreeState<T extends object>(props: TreeProps<T>): TreeState<T
   };
 }
 
-function toggleKey(set: Set<Key>, key: Key): Set<Key> {
-  let res = new Set(set);
-  if (res.has(key)) {
-    res.delete(key);
+function toggleKey<T>(currentExpandedKeys: 'all' | Set<Key>, key: Key, collection: Collection<Node<T>>): Set<Key> {
+  let updatedExpandedKeys: Set<Key>;
+  if (currentExpandedKeys === 'all') {
+    // TODO: would be nice if the collection row information differentiated between childNodes vs childItems
+    // so we didn't have to keep iterating through info, perhaps make the user pass a prop to TreeItem for childItems/hasChildRows even in the static case?
+    updatedExpandedKeys = new Set([...collection].filter(row => {
+      return row.props.childItems || [...collection.getChildren(row.key)].filter(child => child.type === 'item').length > 0;
```

Comment:
related to a previous comment I made, but maybe if we forced the user to tell us if a row has childRows then we wouldn't have to keep checking each of the node's children manually

Reply:
I'm a bit torn here about how much I should be modifying the node information from what is calculated automatically by the RAC collection implementation, especially if I were to add additional properties (in this case adding a new property to mark the item node as having childRows). There is the added difficulty with how these hooks already work with our old collections and thus the node information generated on the old collection side already diverge from those from RAC.

Reply:
ah, i see
maybe instead of in the base collection though, you could do it in TreeCollection? could pre-compute these so you don't have to recompute for every key click? seems like a better time to handle it because there will be idle cycles instead of right when a user tries to interact?

Reply:
sure I could try modifying the old TreeCollection that already exists in stately, but it would mean modifying either the base Node definition that gets used across our collections so that it has a new property or updating TreeCollection here to have a different kind of Node that it uses/returns. Not really sure how we want to handle this tbh and a bit wary of the changes that would need to be made here since we don't really have a tree implementation using our old collections

Reply:
It's almost too bad we don't have some callbacks during collection construction that would allow us to track things. 

I think this is fine for now, it's really only for toggle all

Reply:
probably shouldn't rely on `childItems` here - that might exist but be empty, or maybe in the renderer someone returned null for a certain item. checking the rendered collection will always be right though.

btw it would be more efficient to use `.some` instead of filter here since it will stop as soon as it finds an item.

Reply:
sounds good, I'll update